chia/data_layer/data_store.py

```diff
@@ -821,3 +879,223 @@ async def get_operations(
+                                            None,
+                                            tree_id.hex(),
+                                        )
+                                        internal_nodes += 1
+                                        right_hash = node
+                                else:
+                                    raise RuntimeError(
+                                        f"Did not received expected node. Expected: {node} Received: {hash.hex()}"
+                                    )
+                                if len(stack) > 0:
+                                    node = stack.pop()
+                                else:
+                                    node = None
+                            else:
+                                left_hash = row["left"]
+                                right_hash = row["right"]
+                                left_hash_bytes = hexstr_to_bytes(left_hash)
+                                right_hash_bytes = hexstr_to_bytes(right_hash)
+                                hash = Program.to((left_hash_bytes, right_hash_bytes)).get_tree_hash(
+                                    left_hash_bytes, right_hash_bytes
+                                )
+                                if hash.hex() == node:
+                                    add_to_db_cache[right_hash] = (node, left_hash)
+                                    # At most max_height nodes will be pending to be added to DB.
+                                    assert len(add_to_db_cache) <= 100
+                                else:
+                                    raise RuntimeError(
+                                        f"Did not received expected node. Expected: {node} Received: {hash.hex()}"
+                                    )
+                                stack.append(right_hash)
+                                node = left_hash
+
+                        if node is not None:
+                            raise RuntimeError("Did not download full data.")
+                        await ws.send_str(json.dumps({"type": "close"}))
+
+                await self._insert_root(
+                    bytes32.from_hexstr(root_json["tree_id"]),
+                    bytes32.from_hexstr(root_json["node_hash"]),
+                    Status(root_json["status"]),
+                    root_json["generation"],
+                )
+                return True
+            elif subscription.mode is DownloadMode.HISTORY:
+                async with aiohttp.ClientSession() as session:
+                    async with session.ws_connect(URL) as ws:
+                        request = {
+                            "type": "request_root",
+                            "tree_id": tree_id.hex(),
+                            "node_hash": target_hash.hex(),
+                        }
+                        await ws.send_str(json.dumps(request))
+                        msg = await ws.receive()
```

Comment:
do we need a timeout for all these ws.receive() calls ?

Reply:
added those constants: ```session.ws_connect(URL, timeout=180, heartbeat=60)```

I think at this phrase we don't want failure at all - so I went for a big timeout.